test/core/filters/filter_test.h

```diff
@@ -0,0 +1,201 @@
+  return os << "flags:" << msg.flags()
+            << " payload:" << absl::CEscape(msg.payload()->JoinIntoString());
+}
+
+class FilterTestBase : public ::testing::Test {
+ public:
+  class Call;
+
+  class Channel {
+   private:
+    struct Impl {
+      Impl(std::unique_ptr<ChannelFilter> filter, FilterTestBase* test)
+          : filter(std::move(filter)), test(test) {}
+      size_t initial_arena_size = 1024;
+      MemoryAllocator memory_allocator =
+          ResourceQuota::Default()->memory_quota()->CreateMemoryAllocator(
+              "test");
+      std::unique_ptr<ChannelFilter> filter;
+      FilterTestBase* const test;
+    };
+
+   public:
+    void set_initial_arena_size(size_t size) {
+      impl_->initial_arena_size = size;
+    }
+
+    Call MakeCall();
+
+   private:
+    friend class FilterTestBase;
+    friend class Call;
+
+    explicit Channel(std::unique_ptr<ChannelFilter> filter,
+                     FilterTestBase* test)
+        : impl_(std::make_shared<Impl>(std::move(filter), test)) {}
+
+    std::shared_ptr<Impl> impl_;
+  };
+
+  // One "call" outstanding against this filter.
+  // In reality - this filter is the only thing in the call.
+  // Provides mocks to trap events that happen on the call.
+  class Call {
+   public:
+    ~Call();
+    explicit Call(const Channel& channel);
+
+    Call(const Call&) = delete;
+    Call& operator=(const Call&) = delete;
+
+    // Construct client metadata in the arena of this call.
+    // Optional argument is a list of key/value pairs to add to the metadata.
+    ClientMetadataHandle NewClientMetadata(
+        std::initializer_list<std::pair<absl::string_view, absl::string_view>>
+            init = {});
+    // Construct server metadata in the arena of this call.
+    // Optional argument is a list of key/value pairs to add to the metadata.
+    ServerMetadataHandle NewServerMetadata(
+        std::initializer_list<std::pair<absl::string_view, absl::string_view>>
+            init = {});
+    // Construct a message in the arena of this call.
+    MessageHandle NewMessage(absl::string_view payload = "",
+                             uint32_t flags = 0);
+
+    // Start the call.
+    void Start(ClientMetadataHandle md);
+    // Cancel the call.
+    void Cancel();
+    // Forward server initial metadata through this filter.
+    void ForwardServerInitialMetadata(ServerMetadataHandle md);
+    // Forward a message from client to server through this filter.
+    void ForwardMessageClientToServer(MessageHandle msg);
+    // Forward a message from server to client through this filter.
+    void ForwardMessageServerToClient(MessageHandle msg);
+    // Have the 'next' filter in the chain finish this call and return trailing
+    // metadata.
+    void FinishNextFilter(ServerMetadataHandle md);
+
+    // Mock to trap starting the next filter in the chain.
+    MOCK_METHOD(void, Started, (const ClientMetadata& client_initial_metadata));
```

Comment:
This seems like a bit of a weird use of gmock, since we're not actually mocking out any interface here.  In particular, this seems to push some boilerplate into the tests with the need to use things like `StrictMock<>`, and it requires the test author to understand the specific points at which gmock expectations are checked (generally not until the mock's destruction, which can sometimes be a bit counter-intuitive when reading the code).  It seems like maybe we'd be better off with a bespoke event notification system here, which the tests can check at the specific points where they expect the events to occur, maybe something like the [event queue](https://github.com/grpc/grpc/blob/49b5dfc14f2d3b8723dcc865e93c5a59f2538e4a/test/core/client_channel/lb_policy/lb_policy_test_lib.h#L431) that we use in the LB policy test framework?

Reply:
I'd actually argue that the LB testing framework could use picking up something like this.

We seem to be in the process of designing and building an army of almost-gmock-matchers, each of which will be maintained and behave just a little bit differently.

Instead, I'd prefer to leverage already built and battle tested systems where I can, and save our teams efforts for building RPC systems instead of testing frameworks wherever possible.

Leveraging gmock here means I can use all of the actions built into gmock trivially, and have an already defined playbook for making extensions to the system.